neqo-bin/src/bin/client/main.rs

```diff
@@ -0,0 +1,541 @@
+            "keyupdate" => {
+                self.shared.use_old_http = true;
+                self.key_update = true;
+            }
+            "v2" => {
+                self.shared.use_old_http = true;
+                // Use default version set for this test (which allows compatible vneg.)
+                self.shared.quic_parameters.quic_version.clear();
+            }
+            _ => exit(127),
+        }
+    }
+}
+
+fn get_output_file(
+    url: &Url,
+    output_dir: &Option<PathBuf>,
+    all_paths: &mut Vec<PathBuf>,
+) -> Option<File> {
+    if let Some(ref dir) = output_dir {
+        let mut out_path = dir.clone();
+
+        let url_path = if url.path() == "/" {
+            // If no path is given... call it "root"?
+            "root"
+        } else {
+            // Omit leading slash
+            &url.path()[1..]
+        };
+        out_path.push(url_path);
+
+        if all_paths.contains(&out_path) {
+            eprintln!("duplicate path {}", out_path.display());
+            return None;
+        }
+
+        eprintln!("Saving {url} to {out_path:?}");
+
+        if let Some(parent) = out_path.parent() {
+            create_dir_all(parent).ok()?;
+        }
+
+        let f = OpenOptions::new()
+            .write(true)
+            .create(true)
+            .truncate(true)
+            .open(&out_path)
+            .ok()?;
+
+        all_paths.push(out_path);
+        Some(f)
+    } else {
+        None
+    }
+}
+
+enum Ready {
+    Socket,
+    Timeout,
+}
+
+// Wait for the socket to be readable or the timeout to fire.
+async fn ready(
+    socket: &udp::Socket,
+    mut timeout: Option<&mut Pin<Box<Sleep>>>,
+) -> Result<Ready, io::Error> {
+    let socket_ready = Box::pin(socket.readable()).map_ok(|()| Ready::Socket);
+    let timeout_ready = timeout
+        .as_mut()
+        .map_or(Either::Right(futures::future::pending()), Either::Left)
+        .map(|()| Ok(Ready::Timeout));
+    select(socket_ready, timeout_ready).await.factor_first().0
+}
+
+/// Handles a given task on the provided [`Client`].
+trait Handler {
+    type Client: Client;
+
+    fn handle(&mut self, client: &mut Self::Client) -> Res<bool>;
+    fn maybe_key_update(&mut self, c: &mut Self::Client) -> Res<()>;
+    fn take_token(&mut self) -> Option<ResumptionToken>;
+    fn has_token(&self) -> bool;
+}
+
+/// Network client, e.g. [`neqo_transport::Connection`] or [`neqo_http3::Http3Client`].
+trait Client {
+    fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output;
+    fn close<S>(&mut self, now: Instant, app_error: AppError, msg: S)
+    where
+        S: AsRef<str> + Display;
+    fn is_closed(&self) -> bool;
+}
+
+struct Runner<'a, H: Handler> {
+    local_addr: SocketAddr,
+    socket: &'a mut udp::Socket,
+    client: H::Client,
+    handler: H,
+    timeout: Option<Pin<Box<Sleep>>>,
+    args: &'a Args,
+}
+
+impl<'a, H: Handler> Runner<'a, H> {
+    async fn run(mut self) -> Res<Option<ResumptionToken>> {
```

Comment:
With this commit, the Neqo client implementation has a single `run` function, not one for `http3` and one for `http09`.